crates/rust-analyzer/src/main_loop.rs

```diff
@@ -106,38 +106,41 @@ impl GlobalState {
             );
         };
 
-        let save_registration_options = lsp_types::TextDocumentSaveRegistrationOptions {
-            include_text: Some(false),
-            text_document_registration_options: lsp_types::TextDocumentRegistrationOptions {
-                document_selector: Some(vec![
-                    lsp_types::DocumentFilter {
-                        language: None,
-                        scheme: None,
-                        pattern: Some("**/*.rs".into()),
-                    },
-                    lsp_types::DocumentFilter {
-                        language: None,
-                        scheme: None,
-                        pattern: Some("**/Cargo.toml".into()),
-                    },
-                    lsp_types::DocumentFilter {
-                        language: None,
-                        scheme: None,
-                        pattern: Some("**/Cargo.lock".into()),
-                    },
-                ]),
-            },
-        };
-
-        let registration = lsp_types::Registration {
-            id: "textDocument/didSave".to_string(),
-            method: "textDocument/didSave".to_string(),
-            register_options: Some(serde_json::to_value(save_registration_options).unwrap()),
-        };
-        self.send_request::<lsp_types::request::RegisterCapability>(
-            lsp_types::RegistrationParams { registrations: vec![registration] },
-            |_, _| (),
-        );
+        if self.config.client_caps.dynamic_watched_files {
+            let save_registration_options = lsp_types::TextDocumentSaveRegistrationOptions {
```

Comment:
Why do we only register for didSave if the client is watching the files? Aren't they separate things?

Reply:
If the client sets the `workspace/didChangeWatchedFiles/dynamicRegistration` capability to `false` it means that the server cannot use `client/registerCapability` for `workspace/didChangeWatchedFiles`, doesn't it?

Reply:
I think so... but `textDocument/didSave` isn't controlled by `workspace/didChangeWatchedFiles`. I'm not sure of the relationship here.

Reply:
Yes, you are absolutely right, this is my mistake. I cannot understand how I managed to confuse different capabilities.
I already asked to rollback the PR.

Reply:
The spec is incredibly confusing around this sort of thing :( and you are correct in that we are not checking static/dynamic registrations correctly.

I think that this and the diagnostics bug I probably introduced a few days ago (and the call hierarchy issue I fixed awhile back) highlight the need for more precise tests on our LSP layer.

Reply:
I've fixed it and now use valid capabilities. https://github.com/vsrs/rust-analyzer/commit/87ea3fff6c8d022485cb1dc2b0d19bc7a62dc57c

But I'm not sure that I should open another PR with this fix. I was interested in static/dynamic registrations to try rust-analyzer in Visual Studio. Its LSP implementation turns out to be very... specific. For example it sends `textDocument/didOpen` and `textDocument/documentSymbol` before `initialized` and obviously it does not work. Looks like I'll end up not using default VS LSP at all :(

Reply:
I think it's worth another PR. Maybe we should warn if the client doesn't support it? I'm also not sure how this interacts with our file watching options but that can be discussed in the PR.

Reply:
@vsrs which version of VS is that?

Reply:
@olegtk Microsoft Visual Studio Community 2019, version 16.6.5

Reply:
@vsrs, thanks. VS LSP client is under active development, can you please try VS preview version from 
https://visualstudio.microsoft.com/vs/preview (currently 16.7 Preview5)?

Reply:
@olegtk I tried Preview5. Same problem (`textDocument/didOpen` goes before `initialized`): 
```
C->S: {"jsonrpc":"2.0","id":2,"method":"initialize","params":{...}}
S->C: {"jsonrpc":"2.0","id":2,"result":{...,"serverInfo":{"name":"rust-analyzer","version":"57b4ec4"}}
C->S: {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{...}}
C->S: {"jsonrpc":"2.0","method":"initialized","params":{}}
```

I've tried to catch `textDocument/didOpen` via ILanguageClientMiddleLayer, wait until the server is fully initialized (ILanguageClient.OnServerInitializedAsync called) and resend didOpen notification. It works, but right after `textDocument/didOpen` VS sends `textDocument/documentSymbol` (also before `initialized`) and I did not find a way to catch and delay this request (ILanguageClientMiddleLayer does not work).

Reply:
Hmm, that's not a known issue... can you please point to your ILanguageClient implementation?

Reply:
I don't have a fully functional ILanguageClient, just a quick sketch (enough to reproduce didOpen\initialized on my system):

```cs
    public class ContentDefinition {
        public const string RustContentTypeName = "Rust";

        [Export]
        [Name(RustContentTypeName)]
        [BaseDefinition(CodeRemoteContentDefinition.CodeRemoteContentTypeName)]
        internal static ContentTypeDefinition RustContentTypeDefinition { get; }

        [Export]
        [FileExtension(".rs")]
        [ContentType(RustContentTypeName)]
        internal static FileExtensionToContentTypeDefinition RustFileExtensionDefinition { get; }
    }

    [Export(typeof(ILanguageClient))]
    [ContentType(ContentDefinition.RustContentTypeName)]
    public class Client : ILanguageClient{
        public string Name => "VSRust";
        public IEnumerable<string> ConfigurationSections => null;
        public object InitializationOptions => null;
        public IEnumerable<string> FilesToWatch => null;
        
        public event AsyncEventHandler<EventArgs> StartAsync;
        public event AsyncEventHandler<EventArgs> StopAsync;

        public async Task<Connection> ActivateAsync(CancellationToken token) {
            await Task.Yield();

            var info = new ProcessStartInfo {
                FileName = @"E:\Projects\Rust\rust-analyzer\target\debug\rust-analyzer.exe",
                Arguments = "",
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            info.Environment.Add("RA_WAIT_DBG", "1");
            info.Environment.Add("RA_LOG", "1");
            var process = new Process { StartInfo = info };
            if( process.Start() ) {
                return new Connection(process.StandardOutput.BaseStream, process.StandardInput.BaseStream);
            }

            return null;
        }

        public async Task OnLoadedAsync() {
            var tmp = StartAsync;
            if( tmp != null ) {
                await tmp.InvokeAsync(this, EventArgs.Empty);
            }
        }

        public Task OnServerInitializedAsync() {
            return Task.CompletedTask;
        }

        public Task OnServerInitializeFailedAsync(Exception e) {
            return Task.CompletedTask;
        }
    }
```

Reply:
@olegtk Perhaps it's not the best place to continue discussion. In case you need more details please drop me an email to vit@conrlab.com.

Reply:
Actually you are right, @vsrs, it's a bug in VS LSP client. It's tracked now internally. Thanks!